File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    bool predictTheWinner(vector<int>& A) {
        int n = A.size();
        if (~n & 1) return true;

        vector dp(n, vector<int>(n, -1));
        auto maxDiff = [&](this auto&& maxDiff, int i, int j) -> int {
            int& res = dp[i][j];
            if (res != -1) return res;
            if (i == j) return res = A[i];

            return res = max(A[i] - maxDiff(i + 1, j),
                             A[j] - maxDiff(i, j - 1));
        };

        return maxDiff(0, n - 1) >= 0;
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, A: List[int]) -> bool:
        n = len(A)
        if ~n & 1: return True

        @cache
        def maxDiff(i: int, j: int) -> int:
            if i == j: return A[i]
            return max(A[i] - maxDiff(i + 1, j),
                       A[j] - maxDiff(i, j - 1))

        return maxDiff(0, n - 1) >= 0